core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +60,681 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+	args->a2 = tx_len;
+
+	rxtx->tx_is_mine = false;
+
+	return;
+out:
+	ffa_set_error(args, ret);
+}
+
+static void ffa_mem_relinquish(struct thread_smc_args *args,
+			       struct sp_session *caller_sp,
+			       struct ffa_rxtx  *rxtx)
+{
+	struct sp_mem *smem = NULL;
+	struct ffa_mem_relinquish *mem = rxtx->rx;
+	struct sp_mem_receiver *receiver = NULL;
+	TEE_Result err = FFA_OK;
+	uint32_t exceptions = 0;
+
+	if (!check_rxtx(rxtx)) {
+		ffa_set_error(args, FFA_DENIED);
+		return;
+	}
+
+	exceptions = cpu_spin_lock_xsave(&rxtx->spinlock);
+	smem = sp_mem_get(READ_ONCE(mem->handle));
+
+	if (!smem) {
+		DMSG("Incorrect handle");
+		err = FFA_DENIED;
+		goto err_unlock_rxtwx;
+	}
+
+	if (READ_ONCE(mem->endpoint_count) != 1) {
+		DMSG("Incorrect endpoint count");
+		err = FFA_INVALID_PARAMETERS;
+		goto err_unlock_rxtwx;
+	}
+
+	if (READ_ONCE(mem->endpoint_id_array[0]) != caller_sp->endpoint_id) {
+		DMSG("Incorrect endpoint id");
+		err = FFA_DENIED;
+		goto err_unlock_rxtwx;
+	}
+
+	cpu_spin_unlock_xrestore(&rxtx->spinlock, exceptions);
+
+	receiver = sp_mem_get_receiver(caller_sp->endpoint_id, smem);
+
+	exceptions = cpu_spin_lock_xsave(&mem_ref_lock);
+	if (!receiver->ref_count) {
+		DMSG("To many relinquish requests");
+		err = FFA_DENIED;
+		goto err_unlock_memref;
+	}
+
+	receiver->ref_count--;
+	if (!receiver->ref_count) {
+		cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+		if (sp_unmap_ffa_regions(caller_sp, smem) != TEE_SUCCESS) {
+			DMSG("Failed to unmap region");
+			ffa_set_error(args, FFA_DENIED);
+			return;
+		}
+	} else {
+		cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+	}
+
+	args->a0 = FFA_SUCCESS_32;
+	return;
+
+err_unlock_rxtwx:
+	cpu_spin_unlock_xrestore(&rxtx->spinlock, exceptions);
+	ffa_set_error(args, err);
+	return;
+err_unlock_memref:
+	cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+	ffa_set_error(args, err);
+}
+
+static void zero_mem_region(struct sp_mem *smem, struct sp_session *s)
+{
+	void *addr = NULL;
+	struct sp_ctx *ctx = to_sp_ctx(s->ts_sess.ctx);
+	struct sp_mem_map_region *reg = NULL;
+
+	ts_push_current_session(&s->ts_sess);
+	SLIST_FOREACH(reg, &smem->regions, link) {
+		size_t sz = reg->page_count * SMALL_PAGE_SIZE;
+
+		addr = sp_mem_get_va(&ctx->uctx, reg->page_offset, reg->mobj);
+
+		assert(addr);
+		memset(addr, 0, sz);
+	}
+	ts_pop_current_session();
+}
+
+/*
+ * ffa_mem_reclaim returns false if it couldn't process the reclaim message.
+ * This happens when the memory regions was shared with the OP-TEE endpoint.
+ * After this thread_spmc calls handle_mem_reclaim() to make sure that the
+ * region is reclaimed from the OP-TEE endpoint.
+ */
+bool ffa_mem_reclaim(struct thread_smc_args *args,
+		     struct sp_session *caller_sp)
+{
+	uint64_t handle = 0;
+	uint32_t flags = args->a3;
+	uint32_t endpoint = 0;
+	struct sp_mem *smem = NULL;
+	struct sp_mem_receiver *receiver  = NULL;
+	uint32_t exceptions = 0;
+
+	handle = reg_pair_to_64(args->a2, args->a1);
```

Comment:
suggest to move to where `handle` is defined. For consistency with how `flags` is set.

Reply:
Done